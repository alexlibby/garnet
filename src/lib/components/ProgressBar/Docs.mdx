import { Canvas, Meta } from "@storybook/blocks";
import { ArgTypes } from "@storybook/blocks";
import * as ProgressBarStories from "./ProgressBar.stories";

<Meta of={ProgressBarStories} />

# ProgressBar

ProgressBar is the primary component. It has two possible states.

- [Default](#default)
- [Label Outside](#labeloutside)

## Default

This is the default for our ProgressBar component

<Canvas of={ProgressBarStories.Default} />

## Label Outside

This is set to show the label outside of the ProgressBar field component

<Canvas of={ProgressBarStories.LabelOutside} />

## Properties of component

Below is a list of arguments available for this component:

<ArgTypes of={ProgressBarStories} />
