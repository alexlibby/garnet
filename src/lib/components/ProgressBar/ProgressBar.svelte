<svelte:options customElement="garnet-progressbar" />

<script>
  import { cubicOut } from 'svelte/easing';
  import { tweened } from 'svelte/motion';

  export let progressAmt = 52;
  export let precision = 0;
  export let tweenDuration = 400;
  export let animate = true;
  export let labelInside = true;
  export let labelTextOutside = "This is a test";
  export let easing = cubicOut;
  export let color = '#733635';
  
  const progress = tweened(0, {
    duration: animate ? tweenDuration : 0,
    easing
  });
 
  $: progress.set(Number(progressAmt));
</script>

<div class="garnet-progressbar">
  
  <!-- Adds label inside progressbar -->
  {#if labelInside}
    <div class="labelInsideClass">
      <div class="progressbar" style="width: {$progress}%; background-color: {color}">
        <span class="progress">
          {$progress.toFixed(precision)}%
        </span>
      </div>
    </div>
  {:else}
  <div class="labelOutsideClass">
    <span class="labelTextOutside">{labelTextOutside}</span>
    <span class="progress">
      {$progress.toFixed(precision)}%
    </span>
  </div>
  <div class="progressbar" style="width: {$progress}%; background-color: {color}; height: 10px;"></div>
{/if}
</div>

<style>
  .garnet-progressbar {
    display: flex;
    flex-direction: column;
    font-family: Arial, Helvetica, sans-serif;
  }

  .progress {
    font-size: 14px; 
    line-height: 20px; 
    font-weight: 500;
  }

  .progressbar {
    border-radius: 9999px;
  }

  .labelInsideClass {
    display: flex;
    flex-direction: column;
    font-size: 12px; 
    line-height: 16px;
    font-weight: 500;
    text-align: center;
    padding: 0.5rem;
    border-radius: 9999px;
    color: #000000;
  }

  .labelOutsideClass {
    display: flex;
    justify-content: space-between;
    margin-bottom: 1px;
    color: #000000;
  }

</style>