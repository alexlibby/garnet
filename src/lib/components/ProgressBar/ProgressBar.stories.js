import ProgressBar from "./ProgressBar.svelte";
import { fn } from "@storybook/test";
import { action } from "@storybook/addon-actions";
import { BADGE } from "storybook-addon-badges";

const BADGES = {
  ...BADGE,
  ALPHA: "Alpha",
};

export default {
  title: "Garnet UI Library/Animation Components/ProgressBar",
  component: ProgressBar,
  argTypes: {
    disabled: { control: "boolean" },
    label: { control: "string" },
  },
  on: { change: fn().mockName("on-change") },
  props: {
    disabled: false,
  },
};

export const Default = (args) => ({
  Component: ProgressBar,
  props: {
    ...args,
    color: "red",
    labelTextOutside: "ProgressBar",
  },
  on: { change: fn().mockName("on-change") },
});

Default.parameters = {
  badges: [BADGES.ALPHA],
};

export const LabelOutside = (args) => ({
  Component: ProgressBar,
  props: {
    ...args,
    color: "blue",
    labelInside: false,
    labelTextOutside: "This is a test progress bar",
  },
  on: { change: fn().mockName("on-change") },
});

LabelOutside.parameters = {
  badges: [BADGES.ALPHA],
};
