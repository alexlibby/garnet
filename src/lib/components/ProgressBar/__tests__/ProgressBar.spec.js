import { describe, test, expect } from "vitest";
import { axe } from "vitest-axe";
import * as AxeMatchers from "vitest-axe/matchers";
import { render, screen, fireEvent } from "@testing-library/svelte";
import ProgressBar from "../../ProgressBar/ProgressBar.svelte";

const ProgressBarHTML = `<div class="garnet s-BCNdu6Y8Uyt_"><div class="labelInsideClass s-BCNdu6Y8Uyt_"><div class="progressbar s-BCNdu6Y8Uyt_" style="width: 52%; background-color: red;"><span class="progress s-BCNdu6Y8Uyt_">52%</span></div></div></div>`;
expect.extend(AxeMatchers);

describe("Tests for ProgressBar", () => {
  const props = {
    progress: "52",
    precision: 0,
    tweenDuration: 400,
    animate: true,
    labelInside: true,
    labelTextOutside: "This is a test",
    easing: "cubicOut",
    color: "#733635",
  };

  test("should render properly", () => {
    const result = render(ProgressBar, { props });
    expect(() => result).not.toThrow();
  });

  test("get a snapshot of component", () => {
    // Create a new container for the test and mount component
    const host = document.createElement("div");
    document.body.appendChild(host);
    const instance = new ProgressBar({ target: host, props: props });

    // check that all rendered OK
    expect(instance).toBeTruthy();

    // Take snapshot
    expect(host.innerHTML).toMatchSnapshot();
  });

  test("should demonstrate no issues with accessibility", async () => {
    const render = () => ProgressBarHTML;
    const result = render();

    const results = await axe(result, {
      rules: {
        region: { enabled: false },
      },
    });

    expect(results).toHaveNoViolations();
  });
});
