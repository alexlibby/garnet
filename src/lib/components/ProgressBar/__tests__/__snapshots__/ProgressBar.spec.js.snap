// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Tests for ProgressBar > get a snapshot of component 1`] = `"<div class="garnet svelte-zsgscn"><div class="labelInsideClass svelte-zsgscn"><div class="progressbar svelte-zsgscn" style="width: 0%; background-color: rgb(115, 54, 53);"><span class="progress svelte-zsgscn">0%</span></div></div></div>"`;
